src/middlewared/middlewared/plugins/vm_/events.py

```diff
@@ -0,0 +1,72 @@
+import libvirt
+import threading
+
+from middlewared.service import private, Service
+
+
+class VMService(Service):
+
+    @private
+    def libvirt_connection(self):
+        return self.middleware.call_sync('vm.retrieve_libvirt_connection')
+
+    @private
+    def setup_libvirt_events(self):
+        def callback(conn, dom, event, detail, opaque):
+            """
+            0: 'DEFINED',
+            1: 'UNDEFINED',
+            2: 'STARTED',
+            3: 'SUSPENDED',
+            4: 'RESUMED',
+            5: 'STOPPED',
+            6: 'SHUTDOWN',
+            7: 'PMSUSPENDED'
+            Above is event mapping for internal reference
+            """
+            vms = {f'{d["id"]}_{d["name"]}': d for d in self.middleware.call_sync('vm.query')}
+            if dom.name() not in vms:
+                emit_type = 'REMOVED'
+            elif event == 0:
+                emit_type = 'ADDED'
+            else:
+                emit_type = 'CHANGED'
+
+            vm_state_mapping = {
+                0: 'NOSTATE',
+                1: 'RUNNING',
+                2: 'BLOCKED',
+                3: 'PAUSED',
+                4: 'SHUTDOWN',
+                5: 'SHUTOFF',
+                6: 'CRASHED',
+                7: 'PMSUSPENDED',
+            }
+            try:
+                if event == 1:
+                    if emit_type == 'REMOVED':
+                        state = 'NOSTATE'
+                    else:
+                        # We undefine/define domain numerous times based on if vm has any new changes
+                        # registered, this is going to reflect that
+                        state = 'UPDATING CONFIGURATION'
+                else:
+                    state = vm_state_mapping.get(dom.state()[0], 'UNKNOWN')
+            except libvirt.libvirtError:
+                state = 'UNKNOWN'
+
+            vm_id = dom.name().split('_')[0]
+            if vm_id.isdigit():
+                self.middleware.send_event('vm.query', emit_type, id=int(vm_id), fields={'state': state})
+            else:
+                self.middleware.logger.debug('Received libvirtd event with unknown domain name %s', dom.name())
+
+        def event_loop_execution():
+            while self.libvirt_connection():
+                libvirt.virEventRunDefaultImpl()
```

Comment:
Race condition can occur if `close_libvirt_connection` and then `wait_for_libvirtd` are called during `libvirt.virEventRunDefaultImpl` - this loop will never receive null value of `libvirt_connection`, will never terminate and you'll get two `event_thread` running.

SImple solution is not to use `self.libvirt_connection()` in `setup_libvirt_events`. Just pass connection on which you want to react to events on to `setup_libvirt_events` and run this loop while `connection.isAlive()`.

Reply:
This looks unlikely, however we would still end up with 2 threads if it does happen as whatever happens `wait_for_libvirtd` is going to create another thread and it should as the event handler should be configured with the new connection and not the old one.

Reply:
Actually we also stop libvirt after closing the hypervisor connection and then start it again and then open the connection and start a new thread. It looks highly unlikely that this loop will not yield till then, it yields as soon as the connection is closed. Please advise if you think we should still look after this case. We can perhaps synchronise this in a different manner then

Reply:
You won't need to synchronize anything if you let the event reading thread die with the connection.

Reply:
I just learned something new about objects and how python manages their references. Thank you for pointing this out : )

Reply:
@themylogin when testing this, i see that when we close the connection `isAlive` call fails to correctly determine that the connection is no longer active and results in a traceback which is okay ( would have been better if it correctly told us boolean false ), however we don't want this polluting our logs and we can't ignore it as we would like to log where this happens for a valid/good reason. Suggestions on tackling this ? I have been looking at libvirt source code to see if we could have another flag which might tell us the connection is closed/open but doesn't seem like there is one.

Reply:
And i found a workaround, no worries :)